js/dispatch.ts

```diff
@@ -5,35 +5,86 @@ import * as msg from "gen/msg_generated";
 import * as errors from "./errors";
 import * as util from "./util";
 import { maybePushTrace } from "./trace";
+import { promiseErrorExaminer } from "./promise_util";
 
 let nextCmdId = 0;
 const promiseTable = new Map<number, util.Resolvable<msg.Base>>();
 
-let fireTimers: () => void;
+let fireTimers: () => number;
+let nTasks = 0; // Number of async tasks pending.
+let delay = 0; // Cached return value of fireTimers.
 
-export function setFireTimersCallback(fn: () => void) {
+function eventLoopLog(): void {
+  util.log(`TICK delay ${delay} nTasks ${nTasks}`);
+}
+
+function idle(): boolean {
+  delay = fireTimers();
+  return delay < 0 && nTasks === 0;
+}
+
+export function eventLoop(): boolean {
+  for (;;) {
+    if (idle()) {
+      libdeno.runMicrotasks();
+      if (idle()) {
+        break;
+      }
+    }
+    eventLoopLog();
+    const ui8 = poll(delay);
+    if (ui8 != null) {
+      handleAsyncMsgFromRust(ui8);
+      libdeno.runMicrotasks();
+    }
+  }
+  return promiseErrorExaminer();
```

Comment:
Just curious, does this mean that we will only handle promise uncaught rejects when deno is about to exit?

Reply:
Hmm. I think with this impl deno would not promptly die on uncaught errors after running to the end.

Which one was the original failing test?
(quite busy in the coming weeks but might as well take a look)

Reply:
tests/async_error.ts